node/metered-channel/src/oneshot.rs

```diff
@@ -0,0 +1,353 @@
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Metered variant of oneshot channels to be able to extract delays caused by delayed responses.
+
+use std::{
+	convert::TryFrom,
+	pin::Pin,
+	sync::{
+		atomic::{AtomicU64, AtomicU8, Ordering},
+		Arc,
+	},
+	task::{Context, Poll},
+	time::{Duration, Instant},
+};
+
+use futures::{
+	channel::oneshot::{self, Canceled, Cancellation},
+	future::{Fuse, FusedFuture},
+	prelude::*,
+};
+use futures_timer::Delay;
+use num_enum::{IntoPrimitive, TryFromPrimitive};
+
+/// Provides the reason for termination, or [`Self::Unfinished`] if none exists.
+#[derive(Debug, Clone, Copy, IntoPrimitive, TryFromPrimitive, PartialEq, Eq)]
+#[repr(u8)]
+pub enum Reason {
+	Unfinished = 0,
+	Completion = 1,
+	Cancellation = 2,
+	HardTimeout = 3,
+}
+
+/// A peek into delay and state of a `oneshot` channel.
+#[derive(Debug, Clone, Default)]
+pub struct Meter {
+	inner: Arc<AtomicMeterInner>,
+}
+
+/// Inner meter representation.
+#[derive(Debug, Default)]
+pub struct AtomicMeterInner {
+	/// Delay between first poll and completion or cancellation.
+	first_poll_till_end_in_millis: AtomicU64,
+	/// Determine the reason of completion.
+	reason: AtomicU8,
+}
+
+impl Meter {
+	/// Obtain the duration of a finished or canceled
+	/// `oneshot` channel.
+	/// Returns `None` if the channel still exists.
+	pub fn duration(&self) -> Option<Duration> {
+		if self.reason() == Reason::Unfinished {
+			None
+		} else {
+			let millis = self.inner.first_poll_till_end_in_millis.load(Ordering::Acquire);
+			Some(Duration::from_millis(millis))
+		}
+	}
+
+	/// Obtain the reason to the channel termination.
+	pub fn reason(&self) -> Reason {
+		let reason = self.inner.reason.load(Ordering::Acquire);
+		Reason::try_from(reason).expect(
+			"Input is originally a `Termination` before cast, hence recovery always works. qed",
+		)
+	}
+}
+
+/// Create a new pair of `OneshotMetered{Sender,Receiver}`.
+pub fn channel<T>(
+	name: &'static str,
+	soft_timeout: Duration,
+	hard_timeout: Duration,
+) -> (OneshotMeteredSender<T>, OneshotMeteredReceiver<T>) {
+	let (tx, rx) = oneshot::channel();
+	let shared_meter = Meter::default();
+
+	(
+		OneshotMeteredSender { name, inner: tx, shared_meter: shared_meter.clone() },
+		OneshotMeteredReceiver {
+			name,
+			inner: rx,
+			soft_timeout,
+			hard_timeout,
+			soft_timeout_fut: None,
+			hard_timeout_fut: None,
+			first_poll_timestamp: None,
+			shared_meter,
+		},
+	)
+}
+
+#[allow(missing_docs)]
+#[derive(thiserror::Error, Debug)]
+pub enum Error {
+	#[error(transparent)]
+	Canceled(#[from] Canceled),
+	#[error("No response within {}", Duration::as_secs_f64(.0))]
+	HardTimeout(Duration),
+}
+
+/// Oneshot sender, created by [`channel`].
+#[derive(Debug)]
+pub struct OneshotMeteredSender<T> {
+	name: &'static str,
+	inner: oneshot::Sender<(Instant, T)>,
+	shared_meter: Meter,
```

Comment:
I am missing completely, why the sender needs access to the `Meter`. Actually, I don't understand the purpose of `Meter` to be honest.